pkg/controller/trainingjob_controller.go

```diff
@@ -0,0 +1,260 @@
+	// ExtCli is the extension kubernetes clientset
+	ExtCli extclient.Interface
+	// PaddleCli is a clientset for our own API group
+	PaddleCli paddleclient.Interface
+
+	trainingjobLister paddlelisters.TrainingJobLister
+	trainingjobSynced cache.InformerSynced
+
+	// jobupdater is the collection of job updaters for each training job
+	jobupdater *sync.Map
+
+	// workqueue is a rate limited work queue. This is used to queue work to be
+	// processed instead of performing it as soon as a change happens.
+	workqueue workqueue.RateLimitingInterface
+	// recorder is an event recorder for recording Event resources to the
+	// Kubernetes API.
+	recorder record.EventRecorder
+}
+
+// New returns a trainingjob controller
+func New(
+	kubeCli kubernetes.Interface,
+	extCli extclient.Interface,
+	paddleCli paddleclient.Interface,
+	paddleInformers paddleinformers.SharedInformerFactory) *TrainingJobController {
+	trainingjobInformer := paddleInformers.Paddlepaddle().V1().TrainingJobs()
+	paddlescheme.AddToScheme(scheme.Scheme)
+
+	log.Info("Creating trainingjob event broadcaster")
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(log.Info)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeCli.CoreV1().Events("")})
+	workqueue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "TrainingJob")
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "TrainingJobController"})
+
+	controller := &TrainingJobController{
+		KubeCli:           kubeCli,
+		ExtCli:            extCli,
+		PaddleCli:         paddleCli,
+		trainingjobLister: trainingjobInformer.Lister(),
+		trainingjobSynced: trainingjobInformer.Informer().HasSynced,
+		jobupdater:        new(sync.Map),
+		workqueue:         workqueue,
+		recorder:          recorder,
+	}
+
+	log.Info("Setting up event handlers")
+	trainingjobInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controller.enqueue,
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldTj := oldObj.(*paddlev1.TrainingJob)
+			newTj := newObj.(*paddlev1.TrainingJob)
+			objNs := oldTj.Namespace
+			objName := oldTj.Name
+			if oldTj.ResourceVersion == newTj.ResourceVersion {
+				log.Debug("same resourceversion for training job", objNs, "/", objName, ", skipped")
+				return
+			}
+			log.Debug("resourceversion for training job", objNs, "/", objName, "updated")
+			controller.enqueue(newObj)
+		},
+		DeleteFunc: controller.dequeue,
+	})
+
+	return controller
+}
+
+// Run will set up the event handlers for trainingjob, as well as syncing
+// informer caches and starting workers. It will block until stopCh
+// is closed, at which point it will shutdown the workqueue and wait for
+// workers to finish processing their current work items.
+func (c *TrainingJobController) Run(threadiness int, stopCh <-chan struct{}) error {
+	// TODO add a lock to ensure there is only one controller in the cluster
```

Comment:
Do we need this TODO? Maybe it's the job of the binary (e.g., `cmd/paddle_controller/paddle_controller.go`) to ensure only one instance is running, rather than the library.